torch/csrc/autograd/python_function.cpp

```diff
@@ -56,16 +56,12 @@ PyObject* THPGradientEdgeClass = nullptr;
 // Anonymous namespace for helpful functions used in this file
 namespace {
 
-// Throw a python_error with the PyErr state persisted, so that we
-// don't lose the error state if the GIL is released when we don't
-// have a PyThreadState created beforehand, this is made so that
-// even for pure C++ thread without a pre-created PyThreadState could
-// also capture the correct error message.
-// TODO: This is a temporary approach to allow C++ thread to correctly
-// capture Python Error in autograd, remove this when c10 thread pool
-// allow to do one time initialization.
-// see discussion in https://github.com/pytorch/pytorch/pull/34845
-// Follow up issue: https://github.com/pytorch/pytorch/issues/35006
+// TODO: I'm not sure why we can't remove this function. Since we
+// create a shared PyThreadState upon initialization, we should be able
+// to rely on the engine to persist for us. But, removing this seems to
+// cause the following test to fail:
+// python test/distributed/rpc/test_tensorpipe_agent.py -k test_backward_autograd_engine_error
```

Comment:
You mean removing the use of this function right? Not this function itself?
Is the distributed engine missing a python counterpart that would be persisting the errors in general?
If so I would expect that any error in a hook during backward would hard crash it?

Reply:
> You mean removing the use of this function right? Not this function itself?

It's only used in this file AFAIK, so both?

> Is the distributed engine missing a python counterpart that would be persisting the errors in general?

It holds an instance of the non-dist engine, constructed via get_default_engine, so presumably it would correctly create the right one (haven't verified).

> If so I would expect that any error in a hook during backward would hard crash it?

Good point, I would expect this to be the case (haven't verified).

Reply:
> You mean removing the use of this function right? Not this function itself?

Clarified this comment. 

I've updated the PR to fix the dist engine hook case.